core/src/eval/merge.rs

```diff
@@ -407,47 +408,23 @@ fn merge_fields<'a, C: Cache, I: DoubleEndedIterator<Item = &'a LocIdent> + Clon
     let mut pending_contracts = pending_contracts1.revert_closurize(cache, env_final, env1.clone());
     pending_contracts.extend(pending_contracts2.revert_closurize(cache, env_final, env2.clone()));
 
-    // Annotation aren't used anymore at runtime. We still accumulate them to answer metadata
-    // queries, but we don't need to e.g. closurize or revert them.
-    let mut annot1 = metadata1.annotation;
-    let mut annot2 = metadata2.annotation;
-
-    // If both have type annotations, we arbitrarily choose the first one as the type annotation
-    // for the resulting field. This doesn't make any difference operationally.
-    let typ = match (annot1.typ.take(), annot2.typ.take()) {
-        (Some(ctr1), Some(ctr2)) => {
-            annot1.contracts.push(ctr2);
-            Some(ctr1)
-        }
-        (ty1, ty2) => ty1.or(ty2),
-    };
-
-    let contracts: Vec<_> = annot1
-        .contracts
-        .into_iter()
-        .chain(annot2.contracts)
-        .collect();
-
-    let metadata = FieldMetadata {
-        doc: merge_doc(metadata1.doc, metadata2.doc),
-        annotation: TypeAnnotation { typ, contracts },
-        // If one of the record requires this field, then it musn't be optional. The
-        // resulting field is optional iff both are.
-        opt: metadata1.opt && metadata2.opt,
-        // The resulting field will be suppressed from serialization if either of the fields to be merged is.
-        not_exported: metadata1.not_exported || metadata2.not_exported,
-        priority,
-    };
-
     Ok(Field {
-        metadata,
+        metadata: FieldMetadata {
+            doc: merge_doc(metadata1.doc, metadata2.doc),
+            annotation: Combine::combine(metadata1.annotation, metadata2.annotation),
+            // If one of the record requires this field, then it musn't be optional. The
+            // resulting field is optional iff both are.
+            opt: metadata1.opt && metadata2.opt,
+            not_exported: metadata1.not_exported || metadata2.not_exported,
+            priority,
+        },
         value,
         pending_contracts,
     })
 }
 
 /// Merge two optional documentations.
-fn merge_doc(doc1: Option<String>, doc2: Option<String>) -> Option<String> {
+pub(crate) fn merge_doc(doc1: Option<String>, doc2: Option<String>) -> Option<String> {
```

Comment:
Should we implement this as `Combine`? Although the semantics is a bit arbitrary, we still use it as such. And I believe we have something like `impl<T: Combine> Combine for Option<T>` somewhere?

Reply:
We do. But `String` doesn't implement `Combine` (yet) and as the comment indicates, we're not entirely sure how to combine documentation. So this should probably be another refactor, replacing `Option<String>` by a newtype and implementing an explicit `Combine` instance for it.